### ISSUE TITLE
sly autoloads are erronous

### ISSUE THREAD

**Issue by amno1** [joaotavora/sly#244]
I just installed sly from melpa and am having some troubles with installation. Autoloads are not extracted/installed correctly. The file does not look like one in github repository at all. This is content that got installed on my disk:

`;;; sly-autoloads.el --- automatically extracted autoloads
;;
;;; Code:

(add-to-list 'load-path (directory-file-name
                         (or (file-name-directory #$) (car load-path))))


;;;### (autoloads nil "sly" "sly.el" (0 0 0 0))
;;; Generated autoloads from sly.el

(define-obsolete-variable-alias 'sly-setup-contribs 'sly-contribs "2.3.2")

(defvar sly-contribs '(sly-fancy) "\
A list of contrib packages to load with SLY.")

(autoload 'sly-mode "sly" "\
Minor mode for horizontal SLY functionality.

If called interactively, enable Sly mode if ARG is positive, and
disable it if ARG is zero or negative.  If called from Lisp,
also enable the mode if ARG is omitted or nil, and toggle it
if ARG is `toggle'; disable the mode otherwise.

\(fn &optional ARG)" t nil)

(autoload 'sly-editing-mode "sly" "\
Minor mode for editing `lisp-mode' buffers.

If called interactively, enable Sly-Editing mode if ARG is positive, and
disable it if ARG is zero or negative.  If called from Lisp,
also enable the mode if ARG is omitted or nil, and toggle it
if ARG is `toggle'; disable the mode otherwise.

\(fn &optional ARG)" t nil)

(autoload 'sly "sly" "\
Start a Lisp implementation and connect to it.

  COMMAND designates a the Lisp implementation to start as an
\"inferior\" process to the Emacs process. It is either a
pathname string pathname to a lisp executable, a list (EXECUTABLE
ARGS...), or a symbol indexing
`sly-lisp-implementations'. CODING-SYSTEM is a symbol overriding
`sly-net-coding-system'.

Interactively, both COMMAND and CODING-SYSTEM are nil and the
prefix argument controls the precise behaviour:

- With no prefix arg, try to automatically find a Lisp.  First
  consult `sly-command-switch-to-existing-lisp' and analyse open
  connections to maybe switch to one of those.  If a new lisp is
  to be created, first lookup `sly-lisp-implementations', using
  `sly-default-lisp' as a default strategy.  Then try
  `inferior-lisp-program' if it looks like it points to a valid
  lisp.  Failing that, guess the location of a lisp
  implementation.

- With a positive prefix arg (one C-u), prompt for a command
  string that starts a Lisp implementation.

- With a negative prefix arg (M-- M-x sly, for example) prompt
  for a symbol indexing one of the entries in
  `sly-lisp-implementations'

\(fn &optional COMMAND CODING-SYSTEM INTERACTIVE)" t nil)

(autoload 'sly-connect "sly" "\
Connect to a running Slynk server. Return the connection.
With prefix arg, asks if all connections should be closed
before.

\(fn HOST PORT &optional CODING-SYSTEM INTERACTIVE-P)" t nil)

(autoload 'sly-hyperspec-lookup "sly" "\
A wrapper for `hyperspec-lookup'

\(fn SYMBOL-NAME)" t nil)

(autoload 'sly-info "sly" "\
Read SLY manual

\(fn FILE &optional NODE)" t nil)

(if (or (not (memq 'slime-lisp-mode-hook lisp-mode-hook)) noninteractive (prog1 (y-or-n-p "[sly] SLIME detected in `lisp-mode-hook', which causes keybinding conflicts.\nRemove it for this Emacs session?") (warn "To restore SLIME in this session, customize `lisp-mode-hook'\nand replace `sly-editing-mode' with `slime-lisp-mode-hook'.") (remove-hook 'lisp-mode-hook 'slime-lisp-mode-hook))) (add-hook 'lisp-mode-hook 'sly-editing-mode) (warn "`sly.el' loaded OK. To use SLY, customize `lisp-mode-hook' and\nreplace `slime-lisp-mode-hook' with `sly-editing-mode'."))

(if (fboundp 'register-definition-prefixes) (register-definition-prefixes "sly" '("define-sly-" "inferior-lisp-program" "make-sly-" "sly-")))

;;;***

;;;### (autoloads nil nil ("sly-pkg.el") (0 0 0 0))

;;;***

;; Local Variables:
;; version-control: never
;; no-byte-compile: t
;; no-update-autoloads: t
;; coding: utf-8
;; End:
;;; sly-autoloads.el ends here

I have used use-package to install from melpa. Haven't had time to try directly from git yet.`For some reason 'provide' statement is not generated. It does not setup connection to sbcl correctly either.

**Comment by amno1** [comment]
Cloning from github and manually compiling and importing into emacs worked fine :-).

**Comment by joaotavora** [comment]
I think this is some problem in `use-package`.  If you `package-install`  normally you shouldn't have this problem.  Unfortunately that piece of fine software that is `use-package` is pretty popular. I'll see what I can do.

**Comment by joaotavora** [comment]
https://github.com/slime/slime/pull/512 is SLIME's issue dealing with the same problem.

**Comment by joaotavora** [comment]
@amno1 this _should_ be fixed, can you try again tomorrow or so after MELPA rebuilds SLY?

**Comment by rprimus** [comment]
Wed Jun 26 08:54:25 BST 2019

There is no problem with the `package.el` generated `sly-autoloads.el`.  The generated `autoloads` file (after 3a2d216) is exactly the same as was generated before.

@amno1 To prove this is working, do the following:

```sh
cd <MELPA installation directory for sly>
emacs -Q -L . -l sly-autoloads -f sly
```

The `-l sly-autoloads` reads the generated `sly-autoloads.el` file.  (The git repo `sly-autoloads.el` may have been provided to bootstrap sly during development.)

sly should have no problems starting up - unless there is another problem.

You’ll notice that *most* packages on MELPA do not provide their own `<package>-autoloads.el` file.

**Comment by joaotavora** [comment]
@rprimus, I'm slightly confused. Are you saying that A: problem is solved?  Or are you saying that B: there never was a problem?  Or are you saying that both A: and B: are true?

> The git repo sly-autoloads.el may have been provided to bootstrap sly during development.

Yes.  Another options would be to exclude it from the MELPA recipe, maybe.

**Comment by rprimus** [comment]
Wed Jun 26 09:53:44 BST 2019

@joaotavora I am saying `B` - there never was a problem with the `package.el` generated `sly-autoloads.el`.

>  It does not setup connection to sbcl correctly either.

I believe there may be another problem with the OPs setup/config.

**Comment by joaotavora** [comment]
I think there would eventually be a problem with the previous version of the repo/manual `sly-autoloads.el` and Emacs 27, which is what happened in https://github.com/slime/slime/pull/512

Regardless, I do think that @amno1  may have been confused.  I didn't ask for the `.emacs` recipe we was using, he may have been using `(require 'sly-autoloads)` which makes no sense in a MELPA installation.

**Comment by rprimus** [comment]
Wed Jun 26 10:28:29 BST 2019

Unlike https://github.com/slime/slime/pull/512 and https://github.com/joaotavora/sly-quicklisp/issues/3, the repo/manual `sly-autoloads.el` is being ignored and one is being generated by `package.el`.

**Comment by joaotavora** [comment]
> a good thing

Thanks for the edit, which clears it up. I was a bit confused myself.  Do I need to do anything to the sly contrib repos which also have a repo/manual <sly-contrib>-autoloads.el file?

**Comment by drot** [comment]
This also fixes #216, as for sly contrib repos `sly-macrostep` still doesn't work so I guess the same intervention is needed.

**Comment by rprimus** [comment]
Wed Jun 26 16:04:41 BST 2019

> Do I need to do anything to the sly contrib repos which also have a repo/manual -autoloads.el file?

The contrib repos would need to be `fixed` to work with emacs 27 - along similar lines as slime/slime#512.

I’m trying to figure out what makes `sly` different (in that `package.el` (`package-generate-autoloads`) and `update-directory-autoloads` ignores the repo provided `sly-autoloads.el` file.

**Comment by rprimus** [comment]
Wed Jun 26 16:29:47 BST 2019

Switching discussion to https://github.com/joaotavora/sly-quicklisp/issues/3

**Comment by amno1** [comment]
Hi guys, I am really sorry, I was being busy working, was not even at home last two days. I will try tomorrow to install from the melpa again.

No idea what I would have wrong about my setup, when it comes for sbcl, but connection was setuped fined when I installed from git and compiled with provided makefile. Scbl is in the path, so I guess not much is needed there. I will follow up tomorrow after I tried. Sorry for being innactive. Thanks for the great work on sly and help with the issue.
